### ISSUE TITLE
Introducing sampler-dependent visualization

### ISSUE THREAD

**Issue by contramundum53** [optuna/optuna#3945]
### Motivation

Black-box optimization is inherently heuristic, and in many problems the algorithms don't work well out of the box but needs some hyper-hyper-parameter tuning.

However, Optuna currently don't provide any kind of visualization on the internal states of the samplers, e.g., the posterior distribution of Gaussian processes or the estimated low/high distribution in TPE. 

If we could support such fine-grained visualization of the algorithms, I think it will be of great help for understanding the cause behind (possibly undesirable) behaviors of the heuristics and tuning hyper-hyper-parameters.

### Description

This is one simple example of such visualizations.
![](https://qiita-user-contents.imgix.net/https%3A%2F%2Fqiita-image-store.s3.amazonaws.com%2F0%2F118139%2F79792f9b-76cc-8a56-f94e-0b8c20d3f9f3.gif?ixlib=rb-4.0.0&auto=format&gif-q=60&q=75&w=1400&fit=max&s=e051b727d3d0b2dae5eb587b950ae574)
(Image taken from [here](https://qiita.com/masasora/items/cc2f10cb79f8c0a6bbaa).)

This GIF shows the posterior distribution of the surrogate model (namely Gaussian process) and the computed acquisition function. Such visualization is helpful (and perhaps crucial) for tuning of e.g. the kernel function of Gaussian process.

We could introduce somewhat similar visualizations for TPE and CMA-ES. 

Some potential obstacles are:
* The increased code complexity. Since we need access to the internal state (or possibly local variables inside a function) of the samplers, and since such visualization differs from one sampler to another, we would need extensive discussion for code design.
* Support for higher-dimensional problems. It would be hard or impossible to visualize problems of dimension greater than two or three. Still, I think the behavior in lower dimensional problems do provide important insights.

### Alternatives (optional)

Alternatively, we could simply provide hooks for accessing internal states during optimization (e.g. the `botorch` model used to fit the parameters and function values) and let the user visualize them. However, we would still need to decide which internal states to make public, and the backward compatibility is easily broken.

**Comment by blazespinnaker** [comment]
I think this would be exceedingly useful, not only to better understand the behavior but also to potentially alter it.  For the latter, hooks would be the best place to start, ideally with Chain of Responsibility pattern in order to add/inspect behavior on default implementations.  A use case might be inspecting and altering some of the scoring logic, for example around pruned trials.

For hooks, I'm not sure what the risks are for "backward compatibility" .. can you elaborate?

**Comment by contramundum53** [comment]
Since what internal states a `Sampler` computes belongs to the realm of implementation details, an additional feature / a simple refactoring of a `Sampler` could change the necessary internal states. For example, currently Optuna's `BoTorchSampler` uses `botorch.SingleTaskGP` as the surrogate model, but it can change in the future.

**Comment by contramundum53** [comment]
I made an early prototype of TPE visualization. The following GIFs compare the effect of `consider_magic_clip` option in `TPESampler`. (Other parameters are left to be default.)

|`consider_magic_clip=True` | `consider_magic_clip=False` |
|-|-|
|![anim](https://user-images.githubusercontent.com/24619922/187394627-cb267f92-fce9-4ab1-b65e-73bb0e37534c.gif)| ![anim](https://user-images.githubusercontent.com/24619922/187395830-db31c9e8-aca8-4a88-8772-60cbb3214f3a.gif)|

Note that in the particular case of `TPESampler` with `multivariate=False` (current default), all variables are sampled independently, so the same visualization can be done no matter how many variables you have.

**Comment by contramundum53** [comment]
The code is available [here](https://github.com/contramundum53/optuna-tpe-visualized).

**Comment by contramundum53** [comment]
The problem is that introducing mere possibility for this kind of visualization requires some change in each sampler. I added a logging hook in `TPESampler` to extract the internal states.

**Comment by blazespinnaker** [comment]
For sure,- visualization is an excellent use case of why the **hooks** are very important.   Maybe changing the title or opening a new issue might be a good idea so folks stay focused on the core ideas.

**Comment by nzw0301** [comment]
@contramundum53 Thank you for your explanation. I left my comment before seeing the code, so I expected the changes would be more complicated. Anyway, let me confirm which logic is called _hook_. Do you refer to https://github.com/contramundum53/optuna-tpe-visualized/commit/5ba4bc37e488a2fbc91a74022853744cf76d58a7#diff-1b766d9c12de331da166f3563ab7158597bb9231b985e78b63194f732c3475d9 as the hook?

**Comment by contramundum53** [comment]
@nzw0301 Yes, by _hook_ I meant `algorithm_logger_callback` argument in the above code.

**Comment by contramundum53** [comment]
@blazespinnaker 
Thanks for your opinion!
I think even if we only make visualization public, the implementation would very likely use a hook like the above code. So let me rephrase the issue as follows:

Should we _keep the interface of the hook stable for the future_, or should we _explicitly state that the interface of the hook is unstable and declare that only the visualization feature is stable_?

More specifically, "the interface of the hook" means the content of the dict passed to `algorithm_logger_callback` in the above code.

**Comment by blazespinnaker** [comment]
@contramundum53 your points about backwards compatibility are well considered.  Honest question - can you enumerate some scenarios where it's possible to guarantee forward compatibility visualization but not a stable backing api to the required datasets?

**Comment by contramundum53** [comment]
One example I come up with is the following scenario:
Currently, TPE estimates the `below` distribution and the `above` distribution separately, and both distributions need to be passed to the hook for effective visualization. 
However, a possible algorithmic improvement could utilize density ratio estimation techniques to directly estimate the ratio of `below` and `above` distributions. In this case, the `below` distributions and the `above` distributions are no longer computed, and it will be impossible to maintain the interface of the hook.
The visualization also needs to be updated as such, but note that "visualization feature is stable" means that the visualization continues to be effective, not that the visual output is unchanged.

**Comment by contramundum53** [comment]
Another (less aggressive) example would be:
Currently `BoTorchSampler` supports `CategoricalDistribution` by transforming categorical variables into one-hot vectors and treating them as continuous variables. In this case, it would be convenient for visualization if the hook provides the transformed variables. However, a codefix might stop this treatment and instead directly use `botorch`'s `MixedSingleTaskGP` that can also handle categorical variables. Then, the interface of the hook would also need to change.

**Comment by contramundum53** [comment]
Putting experimental labels would be good, I suppose. (Maybe forever experimental?)

**Comment by blazespinnaker** [comment]
@nzw0301 Who can we reach out to get consensus on this?  Even if the answer is that we just have to fork the library, at least it's an answer.  This could be just a design problem with very little implementation or regression potential - depending on how things play out.

**Comment by nzw0301** [comment]
> Who can we reach out to get consensus on this? 
At least not me.

@contramundum53 do you know the state of this PR? cc: @g-votte.

**Comment by contramundum53** [comment]
@blazespinnaker Sorry for replying late.
We had a (lengthy) discussion about this issue among the core development members. The conclusions are the followings:
* We won't add an API for extracting the internal states of the samplers in Optuna, at least in the next release.
* If you want to visualize the behavior of TPESampler on your function, [the above code](https://github.com/contramundum53/optuna-tpe-visualized) is an example of how you can do it. (You basically need to copy the `Sampler` from Optuna and modify it to expose the internal states in your favorite way, and then use your sampler in `optuna.create_study`.)
* We acknowledge that there are indeed such needs of extracting the internal states of samplers, and we will continue the refactoring discussion of `TPESampler` considering such needs.

The backgrounds behind this is that:
* Although it is helpful for visualizing low-dimensional toy problems, it is still hard to visualize a real-world high-dimensional problem. Thus, such functionality would mainly be used by developers rather than users.
* It is costly to maintain such an API, especially when there are no clear criteria of which internal states to expose.
* The `TPESampler` is already very complicated, and adding a callback to it right now would further increase the complexity and possibly be an obstacle in future refactoring.

**Comment by blazespinnaker** [comment]
Thanks, not sure I agree with the conclusions, but the answer is actually very helpful.  A forking we will go :)

**Comment by contramundum53** [comment]
Now OptunaHub is out and we can add these kind of visualizations there.
